source/common/upstream/eds.cc

```diff
@@ -133,9 +133,11 @@ void EdsClusterImpl::BatchUpdateHelper::updateLocalityEndpoints(
     const envoy::config::endpoint::v3::LocalityLbEndpoints& locality_lb_endpoint,
     PriorityStateManager& priority_state_manager, absl::flat_hash_set<std::string>& all_new_hosts) {
   const auto address = parent_.resolveProtoAddress(lb_endpoint.endpoint().address());
+
   // When the configuration contains duplicate hosts, only the first one will be retained.
+  // An exception is made for internal addresses as passthrough metadata is usually what makes them unique.
   const auto address_as_string = address->asString();
-  if (all_new_hosts.count(address_as_string) > 0) {
+  if (address->envoyInternalAddress() == nullptr && all_new_hosts.count(address_as_string) > 0) {
```

Comment:
I think this could be a bit improved:
```
if (address->type() != Type::EnvoyInternal) {
  const auto address_as_string = address->asString();
  if (all_new_hosts.count(...)) {
    return;
  }
  all_new_hosts.emplace(address_as_string);
}
```

That will avoid counting internal address and unnecessarily creating a string.

Reply:
as a cpp newb.. wouldn't internaladdr != null already avoid the counting? Just double checking.

Reply:
Yeah, but you're still putting internal hosts into a map for no reason.

Reply:
The emplace call has to happen regardless of address->type(), so seems we should always create the string.

Reply:
Didn't see this comment ^ 

Are we certain this map is only used for dedupe? It eventually gets passed here:

https://github.com/envoyproxy/envoy/blob/86eef10a935b9fb3e241c3c998c8d5babb4b3781/source/common/upstream/upstream_impl.cc#L1814-L1824

Reply:
It looks like it's trying to correlate the priority set and the new hosts map, so we should keep them consistent. I wonder if we should make the map be from a pointer instead. Let me look deeper at the usage.